src/jit.c

```diff
@@ -1203,141 +1277,186 @@ static void prepare_translate(struct jit_state *state)
+                  offsetof(riscv_t, X) + 4 * vm_reg_idx);
     return target_reg;
 }
 
-static void ra_load2(struct jit_state *state, int reg_number1, int reg_number2)
-{
-    if (reg_number1 == reg_number2) {
-        vm_reg[1] = vm_reg[0] = ra_load(state, reg_number1);
-        return;
-    }
-    vm_reg[0] = reg_table[reg_number1];
-    vm_reg[1] = reg_table[reg_number2];
-    if (vm_reg[0] == -1) {
-        while (vm_reg[0] == -1 || vm_reg[0] == vm_reg[1]) {
-            count = (count + 1) % n_reg;
-            vm_reg[0] = register_map[count];
-        }
-        store_back_target(state, vm_reg[0]);
-        reg_table[reg_number1] = vm_reg[0];
-        emit_load(state, S32, parameter_reg[0], reg_table[reg_number1],
-                  offsetof(riscv_t, X) + 4 * reg_number1);
-    }
-    if (vm_reg[1] == -1) {
-        while (vm_reg[1] == -1 || vm_reg[0] == vm_reg[1]) {
-            count = (count + 1) % n_reg;
-            vm_reg[1] = register_map[count];
-        }
-        store_back_target(state, vm_reg[1]);
-        reg_table[reg_number2] = vm_reg[1];
-        emit_load(state, S32, parameter_reg[0], reg_table[reg_number2],
-                  offsetof(riscv_t, X) + 4 * reg_number2);
+/* Prevent the host register collision while the first vm register has already
+ * been mapped and the second one is going to be mapped to the same host
+ * register and invoke swapping.
+ */
+static inline int map_vm_reg_reserved(struct jit_state *state,
+                                      int vm_reg_idx,
+                                      int reserved_reg_name)
+{
+    if (reg_table[vm_reg_idx] != -1)
+        return reg_table[vm_reg_idx];
+
+    int target_reg;
+    do {
+        count = (count + 1) % n_host_regs;
+        target_reg = register_map[count].reg_name;
+    } while (target_reg == reserved_reg_name);
+
+    save_reg(state, target_reg);
+    unmap_vm_reg(target_reg);
+    set_reg_table(vm_reg_idx, target_reg);
+    return target_reg;
+}
+
+/* Map two vm registers to the host registers once. The lowest 8 bits encode the
+ * first register and the next higher 8 bits encode the second one.
+ */
+static inline uint32_t map_vm_reg_2(struct jit_state *state,
+                                    int vm_reg_idx1,
+                                    int vm_reg_idx2)
+{
+    int target_reg1, target_reg2;
+    if (vm_reg_idx1 == vm_reg_idx2) {
+        target_reg1 = map_vm_reg(state, vm_reg_idx1);
+        return (((uint32_t) target_reg1) << 8) | ((uint32_t) target_reg1);
     }
+
+    target_reg1 = map_vm_reg(state, vm_reg_idx1);
+    target_reg2 = map_vm_reg_reserved(state, vm_reg_idx2, target_reg1);
+
+    assert(target_reg1 != target_reg2);
+    return (((uint32_t) target_reg2) << 8) | ((uint32_t) target_reg1);
+}
+
+static void ra_load2(struct jit_state *state, int vm_reg_idx1, int vm_reg_idx2)
+{
+    int origin1 = reg_table[vm_reg_idx1], origin2 = reg_table[vm_reg_idx2];
+    uint32_t regs = map_vm_reg_2(state, vm_reg_idx1, vm_reg_idx2);
```

Comment:
I think `map_vm_reg_2` is less efficient because it needs some logic calculaitons to retrive register numbers, maybe this can be modified it to 
```
if (vm_reg_idx1 == vm_reg_idx2) {
   vm_reg[0] = vm_reg[1] = map_vm_reg(state, vm_reg_idx1);
} else {
    vm_reg[0] = map_vm_reg(state, vm_reg_idx1);
    vm_reg[1] = map_vm_reg_reserved(state, vm_reg_idx2, target_reg1);
}

```
and remove `map_vm_reg_2`.